File: garageofcode/other/task3.cc
#include <map>
#include <vector>
#include <math.h>

using namespace std;

int solution(vector<int> A){
    map<int, int> m;

    // Count the number of occurences of each integer
    for(int a : A){
        if(m.count(a)){
            m[a]++;
        }else{
            m[a] = 1;
        }
    }

    int sum = 0;
    for(const auto& [_, num_occurences] : m){
        sum += num_occurences * (num_occurences - 1) / 2;
    }

    return min(sum, 1000000000);
}

int main(){
    vector<int> A = {3, 5, 6, 5, 3, 3};
    printf("%d\n", solution(A));
}

File: garageofcode/other/task3.py
import time
from collections import Counter

import numpy as np
    
def solution(A):
    counter = Counter(A)
    return min(int(sum([n*(n-1) / 2 for n in counter.values()])), 1000000000)


def count_occurences(A):
    d = {}
    for elem in A:
        if elem in d:
            d[elem] += 1
        else:
            d[elem] = 1
    return d

def solution2(A):
    counter = count_occurences(A)
    return min(int(sum([n*(n-1) / 2 for n in counter.values()])), 1000000000)


for _ in range(100):
    test_set = np.random.randint(0, 100, size=0)
    t0 = time.time()
    s1 = solution(test_set)
    t1 = time.time()
    s2 = solution2(test_set)
    t2 = time.time()

    #print(t1 - t0)
    #print(t2 - t1)
    if s1 != s2:
        print("Error!")
        break

print(solution([]))